src/pymor/algorithms/timestepping.py

```diff
@@ -10,227 +10,619 @@
+        M
             The |Operator| M. If `None`, the identity operator is assumed.
         mu
             |Parameter values| for which `operator` and `rhs` are evaluated. The current
             time is added to `mu` with key `t`.
-        num_values
-            The number of returned vectors of the solution trajectory. If `None`, each
-            intermediate vector that is calculated is returned.
+        return_iter
+            Determines the return data, see below.
+        return_times
+            Determines the return data, see below.
 
         Returns
         -------
-        |VectorArray| containing the solution trajectory.
+        U
+            If `return_iter == False` and `return_times == False` (the default), where `U` is a |VectorArray|
+            containing the solution trajectory.
+        (U, t)
+            If `return_iter == False` and `return_times == True`, where `t` is a list of time points corresponding to
+            `U`.
+        iterator
+            If `return_iter == True`, an iterator yielding either `U_n` (if `return_times == False`) or `(U_n, t_n)` in
+            each step.
+        """
+
+        # all checks are delegated to TimeStepperIterator
+        iterator = self.IteratorType(
+                self, t0, t1, U0, A, F=F, M=M, mu=mu, return_iter=return_iter, return_times=return_times)
+        if return_iter:
+            return iterator
+        elif return_times:
+            U = A.source.empty(reserve=self.num_values or 0)
+            t = []
+            for U_n, t_n in iterator:
+                U.append(U_n, remove_from_other=True)
+                t.append(t_n)
+            return U, t
+        else:
+            U = A.source.empty(reserve=self.num_values or 0)
+            for U_n in iterator:
+                U.append(U_n)
+            return U
+
+
+class TimeStepperIterator(BasicObject):
+    """Base class to derive time-stepper iterators from.
+
+    See :meth:`TimeStepper.solve` for a documentation of the init parameters except `stepper`.
+
+    Note: derived classes usually only have to implement :meth:`_step`, and optionally :meth:`_interpolate` if none of
+          the provided interpolations are suitable. Using the interpolation from this base class requires the
+          implementor to store certain data in each step (see :meth:`_interpolate`).
+
+    Parameters
+    ----------
+    stepper
+        The associated :class:`TimeStepper`.
+    """
+
+    def __init__(self, stepper, t0, t1, U0, A, F, M, mu, return_iter, return_times):
+        # check input
+        assert isinstance(stepper, TimeStepper)
+
+        assert isinstance(t0, Number)
+        assert isinstance(t1, Number)
+        assert t1 > t0
+
+        assert isinstance(A, Operator)
+
+        F = F or ZeroOperator(A.source, NumpyVectorSpace(1))
+        assert isinstance(F, (Operator, VectorArray))
+        if isinstance(F, Operator):
+            assert F.source.dim == 1
+            assert F.range == A.range
```

Comment:
I guess this can be removed because it appears below.

Reply:
done by now